### PR TITLE
ExtUI Mesh Leveling Extensions

### PR DESCRIPTION
**Pull request by InsanityAutomation**
Add helper functions and events for Mesh leveling routines.

Still more on my list to add here, but I wanted to get what Im doing in front of @marcio-ao and make sure im not in left field.

Ill be off to read up on the sd card file list and such implementation in the lulzbot branches sometime tomorrow. Example here currently doesnt really give any indication on what to do with that...

### CODE REVIEW

**Comment by thinkyhead** [comment]
Random thoughts….

- Some of the leveling options allow the mesh size to be changed on the fly, or a different region of the bed to be probed. How will the controller acquire the dimensions of the mesh in the firmware? Will the firmware of `ExtUI`-based displays be savvy enough to manage meshes that change dimensions? Do you have access to an `ExtUI`-based controller to work out all the intricacies on both sides of the equation?

- Will `ExtUI`-based displays that don't care about mesh points still produce extra code when compiled? Does our venerable technique of defining empty inline methods reliably produce no function calls in all cases?

- Is it intended for `ExtUI`-based displays to be an integrated part of the firmware with full access to the internal data, or to pull (request) data through an intermediary class, or to be fully passive, just waiting for data pushed from other parts of Marlin?

- Is this the kind of approach we want to continue taking for `ExtUI` going forward: adding direct calls to `ExtUI` everywhere in the code that might potentially want to inform the external display of a state change?
  While we don't have a boatload of SRAM to spare, now would be a good time to look at implementing a proper notification system so that if an object doesn't care about some event, it doesn't even get called.

**Comment by InsanityAutomation** [comment]
> * Some of the leveling options allow the mesh size to be changed on the fly, or a different region of the bed to be probed. How will the controller acquire the dimensions of the mesh in the firmware? Will the firmware of `ExtUI`-based displays be savvy enough to manage meshes that change dimensions? Do you have access to an `ExtUI`-based controller to work out all the intricacies on both sides of the equation?

Currently the screen I am working with has a fixed mesh size in its own firmware, so the example configs will have the define and I intended to use the same define in the ExtUI lib. Its not as flexible as what Marcio is working with.

> * Will `ExtUI`-based displays that don't care about mesh points still produce extra code when compiled? Does our venerable technique of defining empty inline methods reliably produce no function calls in all cases?

We could define a set of config's for ExtUI with what options it supports. 

> * Is it intended for `ExtUI`-based displays to be an integrated part of the firmware with full access to the internal data, or to pull (request) data through an intermediary class, or to be fully passive, just waiting for data pushed from other parts of Marlin?

The intention is to keep the display at arms length, with access to data required (in this case, as im reverse engineering a manufacturer's display without source im fairly fixed on what is required) and enough event notifiers to support common functions. These functions should support most of the mesh adjustment and display I have seen on touchscreens to date. The ExtUI class itself should be fairly completely implemented and maintained so everything is available for screen developers to access. Otherwise we will have difficulty getting any adoption with it as they would be forced to modify the main code anyway.

> * Is this the kind of approach we want to continue taking for `ExtUI` going forward: adding direct calls to `ExtUI` everywhere in the code that might potentially want to inform the external display of a state change?

In my opinion there are constant changes we send continuously (eg temp) and seldom actions we trigger events for. Mesh points changing, filament runout, gcode pause, those need events for the UI to realistically track without the overhead of polling on every scan.


>   While we don't have a boatload of SRAM to spare, now would be a good time to look at implementing a proper notification system so that if an object doesn't care about some event, it doesn't even get called.

Could be done either with a struct of variables for the extui code to set or a set of defines explicit to the extui code at compile time. If a struct of bool is used at least they could be static const.

**Comment by marcio-ao** [comment]
> Ill be off to read up on the sd card file list and such implementation in the lulzbot branches sometime tomorrow. Example here currently doesnt really give any indication on what to do with that...

@InsanityAutomation: Yeah, file handling is not in the examples. The gist of it is that there is a class called `FileList` that encapsulates the underlying Marlin routines. Initially I wanted to make use of the SD class in the GUI, but I found that UltraLCD had special code to handle SDCARD_RATHERRECENTFIRST and SDCARD_SORT_ALPHA, so I made the choice to encapsulates in an intermediate class so the ExtUI did not need to implement the sorting selection.

The ExtUI needs to instantiate a ExtUI::FileList object and use the `seek()`,  `shortFilename()` and `longFilename()` methods to get the file names for the menu. The names will be presented in the sorted order. Here is some pseudocode for drawing the file menu, "ui.addListItem" is a placeholder for however you draw menu items in your UI:

```cpp
void onUpdate() {
   ExtUI::FileList files;
   for(int i = 0; i < files.count(); i++) {
      files.seek(i);
      if(files.isDirectory())
        ui.addMenuItem(">" + files.longFilename());
      else  
        ui.addMenuItem(files.longFilename());
   }
}
```

The ExtUI::FileList object has methods for walking up and down the directory structure, which indirectly calls the underlying SD card methods to do the work. Your UI will probably have a callback for when the user clicks on a file or directory name, as well as when they click on the ".." entry. The handler could instantiate another ExtUI::FileList to convert the index into a filename:

```cpp
void onFileListClick() {
   uint16_t fileIndex = ui.getSelectedFileIndex();
   FileList files;
   files.seek(fileIndex);
   if(files.isDirectory())
      files.changeDir(files.filename());
   else
      ExtUI::print(files.filename());
}

void onFileListDotDot() {
   FileList files;
   file.upDir();
}
```

Anyhow, that's a very brief rundown on how it works. I suppose this ought to be added the example file, if this is clear enough.

Our file list implementation [is here](https://code.alephobjects.com/diffusion/MARLIN/browse/devel/Marlin/src/lcd/extensible_ui/lib/screens/files_screen.cpp), but it is muddled up by all the silly things you have to do for the FTDI display and for implementing pages, so it's not a great reference.

> Is it intended for ExtUI-based displays to be an integrated part of the firmware with full access to the internal data, or to pull (request) data through an intermediary class, or to be fully passive, just waiting for data pushed from other parts of Marlin?

@thinkyhead: When I started to look through UltraLCD, I noticed that the things it did to manipulate the Marlin data structures were far from trivial. My idea was to wrap up that knowledge in ExtUI methods, via an intermediate interface, so that UI designers did not have to re-invent that. So if we stick to that paradigm without exception, all functionality in Marlin would need to be done via the ExtUI API, which means the ExtUI interface would have to be increased many fold. That said, an equally valid argument is that the ExtUI could provide methods only for making a minimum viable UI, suitable for a printer for beginners, while leaving more advanced ExtUIs either make changes via "ExtUI::enqueueCommands_P" or by reaching around and modifying Marlin's internal state as UltraLCD does. Since some panels, I suspect the Due panel being one, operate via GCODE, it seems like that would be a viable option for communicating advanced features and maybe there is a way to leverage that work into ExtUI.

> Will ExtUI-based displays that don't care about mesh points still produce extra code when compiled? Does our venerable technique of defining empty inline methods reliably produce no function calls in all cases?

For the record, my ExtUI implementation manipulates Marlin data through the ExtUI methods only, but I do allow myself to use the ENABLED(FEATURE) macros in my code to eliminate features that are not used.

**Comment by thinkyhead** [comment]
> Since some panels, I suspect the Due panel being one, operate via GCODE, it seems like that would be a viable option for communicating advanced features and maybe there is a way to leverage that work into ExtUI.

G-code is preferred because it requires the least amount of cleverness and has the widest accessibility. My PanelDue patch has been stalled while I've been attending to other things, but I'll get back to it soon. It requires me to modify `cardreader.*` so it can format the file listing output in JSON format. PanelDue also uses a custom G-code to request and receive the current printer state (all the things you would display on a typical status screen) in JSON format. Once these are ready, you might want to consider using the same G-codes. OctoPrint has also shown some interest in these new G-codes.

**Comment by thinkyhead** [comment]
> I do allow myself to use the `ENABLED(FEATURE)` macros in my code

I think that's fine. `ExtUI` is a Marlin component. But it sounds like the device-specific `ExtUI` implementations are also going to need to do this. That's okay, it just makes it harder to share them around without also including all of Marlin to guarantee compatibility. (At least, until we have tagged versions that can be referred-to.)

File: Marlin/src/lcd/extensible_ui/ui_api.cpp [hunk 1]
```diff
@@ -577,6 +577,27 @@ namespace ExtUI {
     return elapsed.value;
   }
 
+   #if HAS_LEVELING
+    bool getLevelingActive() { return planner.leveling_active; }
+    void setLevelingActive(const bool state) { set_bed_leveling_enabled(state) }
+    #if HAS_MESH
+      bool getMeshValid() { return leveling_is_valid(); }
+      float[GRID_MAX_POINTS_X][GRID_MAX_POINTS_Y] getMeshArray() { return Z_VALUES; }
```
Comment (marcio-ao) [inline Marlin/src/lcd/extensible_ui/ui_api.cpp]: What's this? I've never seen anyone return an array from a function. Would you just get a pointer out of this? Isn't the dimensions info meaningless?

Comment (thinkyhead) [reply Marlin/src/lcd/extensible_ui/ui_api.cpp]: I believe C++ will actually type-check this if you're assigning to an array type. The details can be concealed a bit by defining a type for the array:
```cpp
typedef float bed_mesh_t[GRID_MAX_POINTS_X][GRID_MAX_POINTS_Y];
```

Comment (thinkyhead) [reply Marlin/src/lcd/extensible_ui/ui_api.cpp]: Nope! C++ considers this gibberish.

Comment (thinkyhead) [reply Marlin/src/lcd/extensible_ui/ui_api.cpp]: Correct approach is to use an array reference…
```cpp
float (&getMeshArray())[GRID_MAX_POINTS_X][GRID_MAX_POINTS_Y] {
  return Z_VALUES;
}
```
…or…
```cpp
typedef float (&bed_mesh_t)[GRID_MAX_POINTS_X][GRID_MAX_POINTS_Y];
bed_mesh_t getMeshArray() {  return Z_VALUES; }
```

Comment (InsanityAutomation) [reply Marlin/src/lcd/extensible_ui/ui_api.cpp]: Whoops, probably pasted it while in auto pilot not thinking.... Ran through stuff today while installing software updates for the day job so wasn't really paying close enough attention.

**Comment by thinkyhead** [comment]
Rebased to obtain patches that pass Travis…

**Comment by InsanityAutomation** [comment]
In case anybody else stumbles on this, SD printing cant take filename, it needs short filename, so marcio's example above needs to be printFile(files.shortFilename());

**Pull request merged**